### PR TITLE
About docs

### PR DESCRIPTION
**Pull request by oligriffiths**
Added an about doc to explain how broccoli works and the basic concepts.
Comment away with suggestions, it's important we make this guide simple from a nomenclature and concepts POV.

Also added support for a markdown table of contents.

`yarn serve` to test.

### CODE REVIEW

File: src/docs/about.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+---
+title: What is Broccoli?
+---
+
+Broccoli is a simple Javascript build tool, that allows you to configure your build pipeline in Javascript (like you're
```
Comment (stefanpenner) [inline src/docs/about.md]: JavaSciprt

File: src/docs/about.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+those transformations, usually into some distributable form. Typically this will involve things like javascript 
+transformations to allow you to write newer syntax that will work in a browser, to use things like Sass for your CSS,
+etc.
+
+Broccoli.js is different to other build tools. You may be used to tools like [Grunt](https://gruntjs.com/) (a task
+runner), [Gulp](https://gulpjs.com/) (streams and pipes) or [Webpack](https://webpack.js.org/) (a module bundler),
+these all operate on different levels to Broccoli.
+
+Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like
+[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), that
```
Comment (stefanpenner) [inline src/docs/about.md]: webpack, browserify, and many more

File: src/docs/about.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+operate on a set of input files, and emit output files, that is all. Broccoli has no knowledge of the contents of 
+your files, only input and output directories.
+
+## Thinking in Broccoli
+
+Broccoli at its heart is very simple. It's really just a set of connected file transformations, performed by plugins,
```
Comment (stefanpenner) [inline src/docs/about.md]: Maybe something like:

It is really just a set of directories, connected together by a plugin system. This plugin systems allow behavior between two directories to be described. Broccoli then ensures plugins are called at the appropriate time, to ensure the build is succesfull.

File: src/docs/about.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+that manipulate files from one or multiple inputs, into a single output. Each plugin is responsible for processing files
+passed to it in input directories, and writing files to its output directory.
+
+Broccoli handles managing the file-system state, and simply takes input files and passes them as inputs to plugins,
```
Comment (stefanpenner) [inline src/docs/about.md]: broccoli doesn't really care about files, it's simply connects directories and behavior together to achieve what the user wants. Custom behavior is provided by plugins, which are  responsible to read from their input directories, and write to their single output directory.

File: src/docs/about.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+Broccoli at its heart is very simple. It's really just a set of connected file transformations, performed by plugins,
+that manipulate files from one or multiple inputs, into a single output. Each plugin is responsible for processing files
+passed to it in input directories, and writing files to its output directory.
+
+Broccoli handles managing the file-system state, and simply takes input files and passes them as inputs to plugins, 
+and takes the output from each plugin and passes it as an input to the next one. The order in which these operations 
+happen is determined by the build file, a `Brocfile.js` that you're going to learn how to create.
+
+You can think of Broccoli much like a simple programming language, where the output of a function can be passed as the
```
Comment (stefanpenner) [inline src/docs/about.md]: maybe `s/broccoli/broccoli-plugins`

File: src/docs/about.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    uglify(
+        babel('src')
+    ),
+    autoprefixer(
+        sass('src/styles', 'site.scss')
+    )
+]);
+```
+
+In the above, working from inner to outer, a `src` directory is passed to the `babel()` plugin (which will convert our
```
Comment (stefanpenner) [inline src/docs/about.md]: Rather than saying "from inner to outer" might it be clearer to have a number list of steps which sequentializes the pipeline in question? I at-least find "inner to outer" to be hard to follow.

File: src/docs/about.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+There are 3 main concepts to get your head around when using Broccoli:
+
+* [Plugins](#plugins)
+* [Nodes](#nodes)
+* [Trees](#trees)
+
+### Plugins
+
+Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of transforming
```
Comment (stefanpenner) [inline src/docs/about.md]: We spoke about this in todays call, I will review again once you had time to iterate on this section.

**Comment by stefanpenner** [comment]
cc @mike-north could I also get your eyes on this?

File: src/docs/about.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
+---
+title: What is Broccoli?
+---
+
+Broccoli is a simple JavaScript build tool, that allows you to configure your build pipeline in Javascript (like
```
Comment (stefanpenner) [inline src/docs/about.md]: `Javascript` -> `JavaScript`

File: src/docs/about.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  build() {
+    // A plugin can receive single or multiple inputs nodes/directories
+    // The files are available in this.inputPaths[0]/this.inputPaths[1]...
+    // You can do anything with these files that you can do in node
+    // The output of your plugin must write to this.outputPath
+  }
+}
+```
+
+Basic plugins are super simple, anything you can do in node, you can do in the `build()` method. A plugin can
```
Comment (stefanpenner) [inline src/docs/about.md]: I think we should avoid saying phrases like `super simple`. Rather lets show how they are "super simple". For example, "a broccoli-plugin has only 1 purpose to transform the files from this.inputPaths directories to their appropriate output in this.outputPath directory when its build function is invoked."

File: src/docs/about.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+`inputPath` contains files that you can manipulate and write to `this.outputPath`. Broccoli will handle the 
+state of these directories and passing them between plugins.
+
+There is a special case where a `string` is passed as an input to a plugin. When parsing your build pipeline, 
+Broccoli will automatically convert a string input into a
+[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
+directly to its output directory, and optionally allows Broccoli to be notified when files within the input 
+directory change and trigger a rebuild.
+
+### Nodes
```
Comment (stefanpenner) [inline src/docs/about.md]: Do you think we can get away without talking about nodes? I think for the most part, just thinking about broccoli-plugins being connected by broccoli in a graph is sufficient. Yes that implies broccoli-plugins are nodes, but I worry raising nodes and pointing to graph-theory provides less value than cost.

File: src/docs/about.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+transformations to allow you to write newer syntax that will work in a browser, to use a CSS pre-processor like 
+Sass for your CSS, etc.
+
+Broccoli.js is different to other build tools. You may be used to tools like [Grunt](https://gruntjs.com/) (a task
+runner), [Gulp](https://gulpjs.com/) (streams and pipes) or [Webpack](https://webpack.js.org/) (a module bundler),
+these all operate on different levels to Broccoli.
+
+Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like
+[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), 
+[Webpack](https://webpack.js.org/), [Browserify](http://browserify.org/), and many more. These tools operate on a
```
[the remark this answers is not in the document]
Comment (oligriffiths) [reply src/docs/about.md]: Correct, suggestions welcome

File: src/docs/about.md (hunk 10 above)
Comment (oligriffiths) [reply src/docs/about.md]: Yeah I think that's fair.

File: src/docs/about.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+Broccoli works at the file-system level, it provides a JavaScript API to wrap other node tools like
+[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), 
+[Webpack](https://webpack.js.org/), [Browserify](http://browserify.org/), and many more. These tools operate on a
+set of input files, and emit output files, that is all. Broccoli has no knowledge of the contents of your 
+files, only input and output directories.
+
+## Thinking in Broccoli
+
+Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin
```
Comment (mike-north) [inline src/docs/about.md]:
```suggestion
Broccoli consists of a set of directories, connected together by plugins, which describe how files are moved or transformed at each step of the build process.
```

File: src/docs/about.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+system. This plugin systems allow behavior between input and output directories to be described. Broccoli then
```
Comment (mike-north) [inline src/docs/about.md]:
```suggestion
Broccoli
```

File: src/docs/about.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+ensures plugins are called at the appropriate time, to ensure the build is successful, and writes the result to
```
Comment (mike-north) [inline src/docs/about.md]:
```suggestion
ensures plugins are called in the prescribed order, and writes the result to
```

File: src/docs/about.md [hunk 15]
```diff
@@ -0,0 +1,186 @@
+---
+title: What is Broccoli?
+---
+
+Broccoli is a simple JavaScript build tool, that allows you to configure your build pipeline in JavaScript (like
+you're already used to writing), and handles setting up the filesystem state for each transformation that's
```
Comment (mike-north) [inline src/docs/about.md]: I would not necessarily assume that readers understand the distinction between an asset pipeline and a module bundler.

You could say something like

> Broccoli is a build tool, that through allowing developers to compose modular plugins to form an asset pipeline, quickly and reliably combines and transforms source files into deployable static assets.

File: src/docs/about.md [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin
+system. This plugin systems allow behavior between input and output directories to be described. Broccoli then
+ensures plugins are called at the appropriate time, to ensure the build is successful, and writes the result to
+a `target` directory. Each plugin is responsible for processing files passed to it in input directories, and 
+writing files to its output directory.
+
+Broccoli doesn't really care about files, it simply takes source directories and passes them as inputs to 
+plugins, creates an output directory for the plugin to write to, and passes that output directory it as an input
+to the next plugin.
```
Comment (mike-north) [inline src/docs/about.md]: You're taking a lot of time to try and describe something that might be easier to show via a diagram

```
app                                       dist
 └─ src						         	   └─ prod.build-20190206.js
     ├─ index.ts      -->  MyPlugin -->
     └─ other.ts

```

File: src/docs/about.md [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+There are 2 main concepts to get your head around when using Broccoli:
+
+* [Plugins](#plugins)
+* [Trees](#trees)
+
+### Plugins
+
+Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of
+transforming input files into output files. The API of a plugin is pretty simple, all that's required is creating
```
Comment (mike-north) [inline src/docs/about.md]: you may want to be careful of input/output as terminology, since they could be taken to mean pipeline imports or plugin imports.

```suggestion
transforming files at each step of build process. The API of a plugin is pretty simple, all that's required is creating
```

File: src/docs/about.md [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+  build() {
+    // A plugin can receive single or multiple inputs nodes/directories
+    // The files are available in this.inputPaths[0]/this.inputPaths[1]...
+    // You can do anything with these files that you can do in node
+    // The output of your plugin must write to this.outputPath
+  }
+}
+```
+
+A broccoli-plugin has only 1 purpose to transform the files from `this.inputPaths` directories to their output
```
Comment (mike-north) [inline src/docs/about.md]:
```suggestion
A broccoli-plugin has only one purpose: to transform the files from `this.inputPaths` directories to their output
```

File: src/docs/about.md [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+### Trees
+
+Working from the target directory (the final output) back up to the source directories resembles a tree. Think 
+of a piece of Broccoli and you should have a mental model of what a build pipeline looks like.
+
+Broccoli constructs this tree of connected nodes in memory as it parses the `Brocfile`, and sets up the filesystem 
+state for each node, creating an `outputPath` for each node to write to.
+
+You may often encounter the term "tree" when reading plugin READMEs or in tutorials, just remember a tree is a
+connected set of plugins/nodes.
```
Comment (mike-north) [inline src/docs/about.md]: One of the important concepts to get across is that trees can be 

- a string
- something a plugin emits

File: src/docs/about.md [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+
+## Building
+
+Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project. This `js` file sets up
```
Comment (mike-north) [inline src/docs/about.md]: > Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project.
feels like you've said this a few times already

File: src/docs/about.md [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+the build graph, connecting source directories via plugins, and exports a single node whose output will be written
```
Comment (mike-north) [inline src/docs/about.md]: Is it important that consumers of broccoli know about a "build graph"? This sounds like it might be an implementation detail

File: src/content/about.md [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+copyFiles(minified, './build');
+```
+
+The above is all well and good, but notice how we have to handle the state between each operation. Now imagine how
+this scales as our build pipeline grows, adding in Sass compilation, JavaScript transpilation (Babel), 
+fingerprinting, etc. That's lots of temp directories and file state that you have to handle.
+
+Broccoli works by managing of a set of directories, connected together by plugins, which describe how files 
+are moved or transformed at each step of the build process. Broccoli ensures plugins are called in the prescribed
+order, and writes the result to a `target` directory. Each plugin is responsible for processing files passed to
```
Comment (stefanpenner) [inline src/content/about.md]: It might be to low level, but an interesting fact, is if you use the same plugin instance twice, it doesn't result in duplicate work, as plugins auto memoize on a per rebuild basis.

feel free to disregard if this is the wrong place to mention this.

**Comment by stefanpenner** [comment]
Left some minor notes, otherwise LGTM pending @mike-north's sign-off.

Q: should we describe how broccoli-builder can be used in another system? Or is that out of scope for this piece?

File: src/content/about.md (hunk 22 above)
Comment (oligriffiths) [reply src/content/about.md]: I feel like that's too low level

File: src/content/about.md [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+ └─ styles                            │
+     └─ sites.scss --> SassPlugin() ──┘
+```
+
+Broccoli itself doesn't really care about files, it simply takes source directories and passes them as inputs to 
+plugins, creates an output directory for the plugin to write to, and passes that output directory it as an input
+to the next plugin.
+
+Broccoli is configured with a file in the root of your project called `Brocfile.js`. This file defines the build
+pipeline for your application, and is written in plan old JavaScript. The order in which operations happen is
```
Comment (lisaychuang) [inline src/content/about.md]:
```suggestion
pipeline for your application, and is written in plain old JavaScript. The order in which operations happen is
```

File: src/content/about.md [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+    // A plugin can receive single or multiple inputs nodes/directories
+    // The files are available in this.inputPaths[0]/this.inputPaths[1]...
+    // You can do anything with these files that you can do in node
+    // The output of your plugin must write to this.outputPath
+  }
+}
+```
+
+A broccoli-plugin has only one purpose to transform the files from `this.inputPaths` directories to their output
+directory in this.outputPath directory when its `build()` function is invoked. Anything you can do in node,
```
Comment (lisaychuang) [inline src/content/about.md]:
```suggestion
directory in `this.outputPath` directory when its `build()` function is invoked. Anything you can do in node, 
```

File: src/content/about.md [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+  }
+}
+```
+
+A broccoli-plugin has only one purpose to transform the files from `this.inputPaths` directories to their output
+directory in this.outputPath directory when its `build()` function is invoked. Anything you can do in node, 
+you can do in the `build()` method. A plugin can receive one or multiple inputs, and these are available in the 
+`this.inputPaths` array in the order they are provided. `this.inputPaths` contains paths to directories, that are
+the `outputPath` of previous plugins. Each `inputPath` contains files that you can manipulate and write to
+`this.outputPath`. Broccoli will handle the state of these directories and passing them between plugins.
```
Comment (lisaychuang) [inline src/content/about.md]:
```suggestion
`this.outputPath`. Broccoli will handle the state of these directories and take responsibility for passing them between plugins.
```

File: src/content/about.md [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+
+There is a special case where a `string` is passed as an input to a plugin. When parsing your build pipeline, 
+Broccoli will automatically convert a string input into a
+[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
+directly to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
```
Comment (lisaychuang) [inline src/content/about.md]:
```suggestion
to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
```

**Comment by oligriffiths** [comment]
@stefanpenner seems like an advanced doc to me

File: src/content/about.md [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+directory change and trigger a rebuild. You can also manually create an `unwatched` directory from a string by
+using [UnwatchedDir](https://github.com/broccolijs/broccoli-source#new-unwatcheddirdirectorypath-options).
+
+### Trees
+
+Working from the target directory (the final output) back up to the source directories resembles a tree. Think 
+of a piece of Broccoli and you should have a mental model of what a build pipeline looks like.
+
+Broccoli constructs this tree of connected nodes in memory as it parses the `Brocfile`, and sets up the filesystem
```
Comment (lisaychuang) [inline src/content/about.md]: Having read this for the first time, I don't know what a `Brocfile` is by this point.  Maybe call if a configuration file, since you will talk about Brocfile again in ##Building section below.

File: src/content/about.md [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+
+Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project. This `js` file sets up
+the build graph, connecting source directories via plugins, and exports a single node whose output will be written
+to the target directory. Broccoli will then handle wiring up all of the nodes inputs and outputs into a graph 
+(from the end node up to the start nodes), creating temporary directories as it goes, linking inputs to outputs 
+between plugins using symlinks to make them super fast, run the build and invoke the `build()` method on each 
+plugin, and finally resolve all the symlinks and write the files from the final node into the destination build 
+directory.
+
+Confused? Here's an example:
```
Comment (lisaychuang) [inline src/content/about.md]:
```suggestion
Here's an example:
```

File: src/content/about.md [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+mergeTrees(
+    'dir1', => source node, implicitly created when using a string as an input
+    'dir2' => source node, implicitly created when using a string as an input
+)
+module.exports = transformation node with input nodes dir1 and dir2
+```
+
+Thus `module.exports` contains a node that references the two input nodes, and an output path that will contain the
+contents of `dir1` and `dir2` when the `build` command is run. The two input nodes reference two source
+directories, `dir1` and `dir2`.
```
Comment (lisaychuang) [inline src/content/about.md]: Suggest to reverse these 2 sentences.

**Comment by oligriffiths** [comment]
@stefanpenner @mike-north I think we're good to go, addressed the feedback and I think clarified a few other sections, thanks for the reviews.

@lisaychuang Thank you for the review, comments addressed and alterations made.

**Comment by lisaychuang** [comment]
Glad my review was helpful, thanks @oligriffiths !

File: src/content/about.md [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
+to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
+directory change and trigger a rebuild. You can also manually create an `unwatched` directory from a string by
+using [UnwatchedDir](https://github.com/broccolijs/broccoli-source#new-unwatcheddirdirectorypath-options), 
+changes in this directory will not trigger a rebuild.
+
+### Trees
+
+The build pipeline is series of connected plugins, from one or more input directories to a single target directory.
+Working from the target directory (the final output) back up to the source directories resembles a tree. Think of a
```
Comment (stefanpenner) [inline src/content/about.md]: What you have here may be fine, but let me go into more detail so you can make the call if the docs are still appropriate or not.

---

The input structure isn't always tree, for example:

```js
const a = new Foo('foo');
const b = new Bar(a);

merge([a,b]);
```

Although the output of ever plugin is tree like.
Also internally broccoli does cycle breaking, and only processes the plugins as if they were a tree, but connects them as a directed graph.

File: src/content/about.md [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+nodes all the way to the source directories to produce this tree. 
+
+As Broccoli does this, it sets up the filesystem state for each plugin, creating an `outputPath` for each to write
+to. Broccoli normalizes each plugin into what it calls `nodes`, that contain information about the inputs and 
+output of each item in the build pipeline.
+
+You may often encounter the term "tree" when reading plugin READMEs or in tutorials, just remember a tree is a
+connected set of plugins.
+
+## Building
```
Comment (stefanpenner) [inline src/content/about.md]: should we mention that broccoli guarantees to only be building 1 plugin at a time per pipeline.

File: src/content/getting-started.md [hunk 32]
```diff
@@ -40,7 +40,9 @@ if required, however installing within the project allows for `yarn` script comm
 Next create a `Brocfile.js` file and an `src` folder for your source code.
 
 ```shell
-$ echo "module.exports = 'src'" > Brocfile.js && mkdir src
+$ echo "module.exports = 'src'" > Brocfile.js
```
Comment (stefanpenner) [inline src/content/getting-started.md]: Should this be updated, considering the recent improvements re:

* esm
* function export

File: src/content/about.md (hunk 30 above)
Comment (oligriffiths) [reply src/content/about.md]: Sure, but I think from a mental model, it's still useful to think of it as a tree.
```
  a
/  \
|   b
\  /
 c
```
The above is more of a shrub than a tree



If you feel strongly about this, please say so, but I think the overall mental model is useful to think of like a tree.

Comment (stefanpenner) [reply src/content/about.md]: Nah, I'm onboard. I just wanted to make sure this detail doesn't detract from the mental model. I do agree with you, I think it's fine. Saying "broccoli trees are actually directed graphs" likely just confuses the story, for little gain.

File: src/content/about.md (hunk 31 above)
Comment (oligriffiths) [reply src/content/about.md]: updated

File: src/content/about.md [hunk 33]
```diff
@@ -91,60 +91,70 @@ app                                                       [target]
...
+Finally, the `export default () => ` line exports a function that Broccoli will invoke with an `options` object,
+that contains an `env` property to indicate the environment. This is set from the `--environment` CLI argument and
+defaults to `development`.
+
 This should all be fairly familiar to you if you've ever written JavaScript (or any programming language for that
 matter) before, it's just inputs and output.
 
+Note: [ES Modules](https://nodejs.org/api/esm.html) syntax is supported by Broccoli in the current master. Prior
+to this, standard CommonJs syntax with `require` and `module.exports` should be used.
+
```
Comment (mike-north) [inline src/content/about.md]: out of curiosity, does this basically just mean you're providing default exports now?

Comment (oligriffiths) [reply src/content/about.md]: The standard practice will be to use `export default` from the Brocfile. Is that what you mean?
